comprl/server/managers.py

```diff
@@ -282,20 +320,132 @@ def remove(self, player_id: PlayerID) -> None:
         Args:
             player_id (PlayerID): The ID of the player to be removed.
         """
-        self.queue = [p for p in self.queue if (p != player_id)]
+        self._queue = [
+            (p_id, u_id, mu, sigma, time)
+            for p_id, u_id, mu, sigma, time in self._queue
+            if (p_id != player_id)
+        ]
 
     def update(self):
         """
         Updates the matchmaking manager.
         """
-        pass
+
+        if len(self._queue) < self._min_players_waiting():
+            return
+
+        for i in range(len(self._queue)):
+            for j in range(i + 1, len(self._queue)):
+                # try to match all players against each other
+                if self._try_start_game(self._queue[i], self._queue[j]):
+                    return
+        return
+
+    def _min_players_waiting(self) -> int:
+        """
+        Returns the minimum number of players that need to be waiting in the queue.
+
+        Returns:
+            int: The minimum number of players.
+        """
+        return int(
+            len(self.player_manager.auth_players) * self._PERCENTAGE_MIN_PLAYERS_WAITING
+        )
+
+    def _try_start_game(self, player1: QueuePlayer, player2: QueuePlayer) -> bool:
+        """
+        Tries to start a game with the given players.
+
+        Args:
+            player1 (QueuePlayer): The first player.
+            player2 (QueuePlayer): The second player.
+
+        Returns:
+            bool: True if the game was started, False otherwise.
+        """
+        player1_id, user1_id, _, _, _ = player1
+        player2_id, user2_id, _, _, _ = player2
+        match_quality = self._rate_match_quality(player1, player2)
+        # prevent the user from playing against himself
+        if user1_id == user2_id:
+            return False
+
+        if match_quality > self._MATCH_QUALITY_THRESHOLD:
+            # match the players. We could search for best match but using the first adds
+            # a bit of diversity and the players in front of the queue are waiting
+            # longer, so its fairer for them.
+
+            # remove players from queue
+            self.remove(player1_id)
+            self.remove(player2_id)
+
+            players = [
+                self.player_manager.get_player_by_id(player1_id),
+                self.player_manager.get_player_by_id(player2_id),
+            ]
+
+            filtered_players = [player for player in players if player is not None]
```

Comment:
I think filtering the players here is not the best solution, because this could potentially lead to starting a game with only one ore no players. It would be better to check if the players are `None` and don't match them in this case (remove them from matchmaking and player manager)

Reply:
You are right. this should never happen, but it's better to catch this error. I changed it.